source/extensions/filters/network/ratelimit/ratelimit.cc

```diff
@@ -18,16 +23,44 @@ namespace RateLimitFilter {
 Config::Config(const envoy::extensions::filters::network::ratelimit::v3::RateLimit& config,
                Stats::Scope& scope, Runtime::Loader& runtime)
     : domain_(config.domain()), stats_(generateStats(config.stat_prefix(), scope)),
-      runtime_(runtime), failure_mode_deny_(config.failure_mode_deny()) {
+      runtime_(runtime), failure_mode_deny_(config.failure_mode_deny()),
+      request_headers_(Http::RequestHeaderMapImpl::create()),
+      response_headers_(Http::ResponseHeaderMapImpl::create()),
+      response_trailers_(Http::ResponseTrailerMapImpl::create()) {
   for (const auto& descriptor : config.descriptors()) {
     RateLimit::Descriptor new_descriptor;
     for (const auto& entry : descriptor.entries()) {
       new_descriptor.entries_.push_back({entry.key(), entry.value()});
+      substitution_formatters_.push_back(
+          std::make_unique<Formatter::FormatterImpl>(entry.value(), false));
     }
     descriptors_.push_back(new_descriptor);
   }
 }
 
+void Config::applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
```

Comment:
Per other comments, this function needs to return the final descriptors, not modify the config inline, which is not thread safe. Also, you don't need to pass original_descriptors, you already have it within the config.